### DISCUSSION TITLE
Do you see a possibility to use the Migrator within Azure Devops Pipelines?

### DISCUSSION THREAD

**Discussion by Mindia620** [Zerg00s/FlowPowerAppsMigrator#49] [Q&A] [answered]
As we move our work over from manually importing releases to Azure DevOps, so we would like to use the Migrator in there...
Is this even possible?

**Comment by Zerg00s** [comment, marked as answer]
Hi @Mindia620 ,

Yes, it is possible to make this migrator work in Azure DevOps pipeline. But that might take several hours of debugging. I would first make sure it works on your local computer, to save time. And then, move it to the pipeline. Overall, the the migrator is just 3 scripts combined together:
- GenerateInitialMapping.ps1
- CompleteResourceMapping.ps1
- ConvertPackage.ps1

The `RunAllScripts.ps1` file combines these files and adds some fluff for user-friendly executions.

Although I haven't tried it myself, there are a few things you can do:

## Use Windows Server VM for the pipeline. Not the Linux one.

## Install and import the PnP.PowerShell.Online module via the pipeline
You can add a step in the ADO pipeline that installs the legacy PowerShell module with
Install-Module PnP.PowerShell.Online -force -allowclobber

## Update the RunAllScripts.ps1 file:

It is fine to modify the RunAllScripts.ps1  file to match your needs. You can also choose to abandon RunAllScripts.ps1 completely. It does not really have any migration logic in it.
- Comment out the `Import-Module (Get-ChildItem -Recurse -Filter "*.psd1").FullName -DisableNameChecking` line
- Comment out the `. .\MISC\PS-Forms.ps1` linme
- Remove this code snippet:
```
if ($Migration.MIGRATE_LISTS -like "true" -or 
    $Migration.MIGRATE_LISTS -like "yes" -or
    $Migration.MIGRATE_LISTS -like "1"
) {
    $Migration.MIGRATE_LISTS = $true
}
else {
    $Migration.MIGRATE_LISTS = $false
}
$MIGRATE_LISTS = $Migration.MIGRATE_LISTS


if ($Migration.CLEAR_CREDENTIALS_CACHE -like "true" -or 
    $Migration.CLEAR_CREDENTIALS_CACHE -like "yes" -or
    $Migration.CLEAR_CREDENTIALS_CACHE -like "1"
) {
    $Migration.CLEAR_CREDENTIALS_CACHE = $true
}
else {
    $Migration.CLEAR_CREDENTIALS_CACHE = $false
}
```
- Update all lines with `Connect-PnPOnline` to use different authentication method. So, instead of 
`Connect-PnPOnline -Url $TARGET_SITE_URL -SPOManagementShell -ClearTokenCache -WarningAction Ignore`
or 
`Connect-PnPOnline -Url $TARGET_SITE_URL -UseWebLogin -WarningAction Ignore`

Update it to use the Azure App Registration or SharePoint-only app. for example:
`Connect-PnPOnline -Url $TARGET_SITE_URL -ClientId $appId -ClientSecret $appSecret `

## Automate export and import of Power Platform solutions with

### Create Azure App registration
- Create an azure app registration for your Power App CLI. The App registration does not need any permissions whatsoever. 
- Create an App secret. 
- Create Application User in the Power Platform Org Admin center. Read on how to do it here:
   https://learn.microsoft.com/en-us/power-platform/admin/manage-application-users#create-an-application-user
   https://learn.microsoft.com/en-us/power-platform/developer/cli/reference/auth#named-create-with-service-principal

Then you will be able to use cli to export and import solutions in your Azure DevOps pipeline

### Install the MS PP CLI:
`pac install latest`

[Authenticate to an environment](https://docs.microsoft.com/en-us/powerapps/developer/component-framework/import-custom-controls#connecting-to-your-environment)

$clientSecret = 'XXXXavb-j'
pac auth create --url "https://orge1bfab21.crm3.dynamics.com/" --name DemoName --applicationId c2eed917-8d34-4583-8396-564da4dd5851 --clientSecret $clientSecret --tenant 872e58d1-c7a9-4519-9595-600305d796a7


### Export Solution
	`pac solution export --path "C:\Solution.zip" --name AppSample --managed true --include general`
	
### Run the Flow and Apps Migrator here

### Import the converted Solution
	`pac solution import --path c:\Solution.zip`

**Comment by Mindia620** [comment]
Thank you very much for your detailed answer. We will look into doing that.
